pydantic/types.py

```diff
@@ -626,3 +634,238 @@ def display(self) -> str:
+        """
+        Return 6-digit hexadecimal value from the 3-digit format, fallback to original value
+        """
+        if len(value) != 3:
+            return value
+        return '{0}{0}{1}{1}{2}{2}'.format(value[0], value[1], value[2])
+
+    @staticmethod
+    def reduce_6_digit_hex(value: str) -> str:
+        """
+        Return 3-digit hexadecimal value from 6-digit, fallback to original value
+        """
+        if len(value) != 6:
+            return value
+        a, b, c, d, e, f = value
+        return '{a}{c}{e}'.format(a=a, c=c, e=e) if (a == b and c == d and e == f) else value
+
+    def _parse_rgb_str(self, value: str) -> Optional[str]:
+        """
+        Get name of the color by its RGB/RGBA string
+        """
+        rgba = self._rgb_str_to_tuple(value)
+        return self._parse_tuple(rgba)
+
+    def _parse_hex_str(self, value: str) -> Optional[str]:
+        """
+        Get name of the color by its hexadecimal string
+        """
+        if value.startswith('#'):
+            pure_hex = value[1:]
+        elif value.startswith('0x'):
+            pure_hex = value[2:]
+        else:
+            pure_hex = value
+
+        if len(pure_hex) == 3:
+            pure_hex = self.expand_3_digit_hex(pure_hex)
+
+        return colors.BY_HEX.get(pure_hex)
+
+    def _parse_color(self) -> None:
+        """
+        Main logic of color parsing
+        """
+        if isinstance(self._original, tuple):
+            self._color_match = self._parse_tuple(self._original)
+
+        elif isinstance(self._original, str):
+            color = self._original.lower()
+
+            # rgb/rgba string
+            self._color_match = self._parse_rgb_str(color)
+            if self._color_match is not None:
+                return
+
+            # hex string
+            self._color_match = self._parse_hex_str(color)
+            if self._color_match is not None:
+                return
+
+            # named colour
+            if color in colors.BY_NAME:
+                self._color_match = color
+                return
+
+    def _get_color_or_raise(self) -> None:
+        """
+        Raise error if color name not found
+        """
+        if self._color_match is None:
+            raise errors.ColorError()
+
+    def original(self) -> ColorType:
+        """
+        Return original value passed to the model
+        """
+        return self._original
+
+    def as_hex(self, prefix: str = '0x', reduce: bool = True) -> str:
+        """
+        Return hexadecimal value of the color
+
+        Try reduce hex code to 3-digit format, fallback to 6-digit.
+        """
+        self._get_color_or_raise()
+        hexadecimal, _ = colors.BY_NAME[self._color_match or '']
+        return '{}{}'.format(prefix, self.reduce_6_digit_hex(hexadecimal) if reduce else hexadecimal)
+
+    def as_tuple(self, alpha: bool = False) -> AnyRGBType:
+        """
+        Return RGB or RGBA tuple
+        """
+        self._get_color_or_raise()
```

Comment:
I think thse methods can assume that `self._get_color_or_raise()` has already been called?